mtt/src/mtt.rs

```diff
@@ -0,0 +1,688 @@
+    NonConfidential1G(Mtt1GEntry<'a, NonConf>),
+    // The entire 1GB range containing the address is confidential.
+    // 1GB ranges are mapped using 64MB subranges, and by convention, the invariant is
+    // that each of the 16 consecutive entries in the L2 table have the same type.
+    Confidential1G(Mtt1GEntry<'a, Conf>),
+    // The 64-MB range has been partitioned into 2MB regions, and each sub-region can be
+    // confidential or non-confidential
+    Mixed2M(Mtt64MEntry<'a>),
+    // The 64-MB range has been partitioned into 16K regions of 4KB size.
+    // Each sub-region can be confidential or non-confidential
+    L1Entry(MttL1Entry<'a>),
+}
+
+struct MttL2Directory {
+    mtt_l2_base: &'static mut [u64],
+}
+
+impl MttL2Directory {
+    fn get_memory_type(addr: SupervisorPageAddr, l2_entry: &MttL2Entry) -> MttMemoryType {
+        use MttL2Entry::*;
+        match l2_entry {
+            Confidential1G(_) => MttMemoryType::Confidential,
+            NonConfidential1G(_) => MttMemoryType::NonConfidential,
+            Mixed2M(mixed_2m_entry) => {
+                if mixed_2m_entry.is_conf(addr) {
+                    MttMemoryType::Confidential
+                } else {
+                    MttMemoryType::NonConfidential
+                }
+            }
+            L1Entry(l1_entry) => {
+                if l1_entry.is_conf(addr) {
+                    MttMemoryType::Confidential
+                } else {
+                    MttMemoryType::NonConfidential
+                }
+            }
+        }
+    }
+
+    fn set_memory_type(addr: SupervisorPageAddr, l2_entry: MttL2Entry, memory_type: MttMemoryType) {
+        use MttL2Entry::*;
+        match l2_entry {
+            Confidential1G(conf_1g) => conf_1g.set_memory_type(memory_type),
+            NonConfidential1G(nc_1g) => nc_1g.set_memory_type(memory_type),
+            Mixed2M(mut mixed_2m_entry) => mixed_2m_entry.set_memory_type(addr, memory_type),
+            L1Entry(mut l1_entry) => l1_entry.set_memory_type(addr, memory_type),
+        }
+    }
+
+    // Returns the index of the MTT L2 entry for the physical address
+    fn get_mtt_index(phys_addr: SupervisorPageAddr) -> usize {
+        let addr = phys_addr.bits();
+        // Bits 45:26
+        ((addr & 0x3fff_fc00_0000) >> 26) as usize
+    }
+
+    fn entry_for_addr(&mut self, phys_addr: SupervisorPageAddr) -> Result<MttL2Entry> {
+        use MttL2Entry::*;
+        let mtt_index = Self::get_mtt_index(phys_addr);
+        let value = self.mtt_l2_base[mtt_index];
+        if (value & L2_ZERO_MASK) != 0 {
+            return Err(InvalidL2Entry);
+        }
+        let mtt_l2_type = (value & L2_TYPE_MASK) >> L2_TYPE_SHIFT;
+
+        match mtt_l2_type {
+            // INFO must be 0 for 1GB entries
+            L2_1G_NC_TYPE if (value & L2_INFO_MASK) == 0 => Ok(NonConfidential1G(
+                Mtt1GEntry::new_nc(&mut self.mtt_l2_base[mtt_index..=mtt_index + 15]),
```

Comment:
hm, is this right? don't we need to start from the base of the 1G range in case `phys_addr` is somewhere in the middle?